test/message/promise_unhandled_reject.js

```diff
@@ -0,0 +1,53 @@
+// Flags: --expose-gc
+'use strict';
+
+const common = require('../common');
+// TODO: Remove when making this the default.
+process.env.NODE_UNHANDLED_REJECTION = 'ERROR_ON_GC';
+
+// Verify that unhandled rejections cause `process.exit` to be called with exit
+// code 1 if they get garbage collected. Any promise that is handled later on
+// should not cause this.
+
+let p1 = new Promise((res, rej) => {
+  throw new Error('One');
+});
+
+// eslint-disable-next-line no-unused-vars
+let p2 = new Promise((res, rej) => {
+  throw new Error('Two');
+});
+
+const p3 = Promise.reject(new Error('Three'));
+
+global.gc();
+
+process.nextTick(() => {
+  console.log('First GC did not collect it.');
+  global.gc();
+  // Should not trigger an error.
+  p1.catch(() => {});
+  p1 = null;
+  setImmediate(() => {
+    console.log('Second GC did not collect it.');
+    setTimeout(() => {
+      // Should still trigger a warning.
+      p3.catch(() => {});
+    }, 1000);
+    global.gc();
+    setImmediate(() => {
+      console.log('Third GC did not collect it.');
+      // Free the reference and exit on GC
+      p2 = null;
```

Comment:
Does this need to be set to null? Shouldn't the GC be able to collect that object regardless?

Reply:
How? There is a reference to it from the module scope

Reply:
If a variable is not used, V8 is able to optimize it out unless you do things like eval()

Reply:
By using `p2` here you force it to be context-allocated, which makes it live longer than necessary, actually :P

Reply:
True, that was an oversight.